# SalesInsight PY — Análise e Visualização de Dados de Vendas

<div style="
    border: 1px solid #d0d7de;
    border-radius: 8px;
    padding: 12px 16px;
    margin-bottom: 20px;
">
<table style="width: 100%; border: none;">
  <tr>
    <!-- Logo SENAI -->
    <td style="
        width: 20%;
        text-align: left;
        vertical-align: middle;
        border: none;
    ">
      <img
        src="../images/senai_logo.jpeg"
        alt="SENAI"
        style="height: 55px;">
    </td>
    <!-- Informações pessoais -->
    <td style="
        width: 60%;
        text-align: center;
        vertical-align: middle;
        border: none;
        line-height: 1.6;
    ">
      <strong style="font-size: 16px;">
        Bruno Miguel Corrêa
      </strong>
      <br>
      <span style="font-size: 14px;">
        Desenvolvimento de IA para Análise Preditiva
      </span>
      <br><br>
      <!-- GitHub -->
      <a
        href="https://github.com/SEU_USUARIO"
        target="_blank"
        style="
            text-decoration: none;
            margin-right: 20px;
        ">
        <img            src="https://cdn.simpleicons.org/github/181717"
          width="24"
          height="24"
          alt="GitHub">
      </a>
      <!-- LinkedIn -->
      <a
        href="https://www.linkedin.com/in/SEU_USUARIO/"
        target="_blank"
        style="text-decoration: none;">
        <img         src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg"
          width="24"
          height="24"
          alt="LinkedIn">
      </a>
    </td>
    <!-- Logo do Programa -->
    <td style="
        width: 20%;
        text-align: right;
        vertical-align: middle;
        border: none;
    ">
      <img
        src="../images/sc_tech_logo.jpeg"
        alt="Programa"
        style="height: 55px;">
    </td>
  </tr>
</table>

</div>

### Sobre o Projeto

O `SalesInsight PY` é um projeto de análise de dados de vendas desenvolvido em Python com o objetivo de carregar, inspecionar, limpar, transformar, agregar e visualizar ados comerciais.

O projeto utiliza `Pandas`, `NumPy`, `Matplotlib` e consolida os contreúdos desenvolvidos em aula.

### Ambiente e Dependências

As bibliotecas utilizadas ao longo do projetosão importadas de forma centralizada nesta seção. As versões são apresnetadas para facilitar a reprodutibilidade do ambiente de execução.

In [1]:
import platform
from pathlib import Path

import pandas as pd
import numpy as np
import re

from sales_insight.dicionario_dados import DICIONARIO_DADOS

print("=" * 6 + " Ambiente de Execução " + "=" * 6)
print(f"{"Python":<12}: {platform.python_version()}")
print(f"{"Pandas":<12}: {pd.__version__}")
print(f"{"NumPy":<12}: {np.__version__}")
print(f"{"Regex":<12}: {re.__version__}")

====== Ambiente de Execução ======
Python      : 3.13.14
Pandas      : 3.0.5
NumPy       : 2.5.1
Regex       : 2.2.1


## 1. Carregamento e Inspeção Inicial dos Dados

### O Dataset `vendas.csv`

**Geração:**

- O dataset tem como ponto de partida o [**E-commerce Analytics Dataset Brazil**](https://www.kaggle.com/datasets/joocarlosjr/e-commerce-analytics-dataset-brazil), disponibilizado publicamente no Kaggle.
- Os arquivos CSV da fonte original foram integrados para formar uma base consolidada de vendas.
- A base foi posteriormente adaptada e enriquecida com registros sintéticos para ampliar sua cobertura temporal e tornar a distribuição das vendas menos uniforme entre os períodos.
- Novos clientes e produtos foram incorporados preservando relações coerentes entre atributos como cliente, cidade, estado, região, produto e categoria.
- Algumas variáveis da fonte original foram removidas por não contribuírem diretamente para os objetivos analíticos do projeto.
- Foram adicionadas informações comerciais e logísticas, como descontos, previsão de entrega e data de entrega, utilizando regras controladas de geração.
- Foram introduzidas propositalmente inconsistências e ruídos nos dados para possibilitar as etapas de limpeza, tratamento de datas, padronização textual e uso de expressões regulares previstas no projeto.


### `RF01` - Carregar o Dataset de Vendas

O dataset `vendas.csv`, está armazenado em `data/raw`. O arquivo será lido com _Pandas_ e convertido em um `DataFrame`, que servirá como base para as etapas seguintes de inspeção, limpeza e análise.

In [2]:
df_origin = pd.read_csv("../data/raw/vendas.csv")
df = df_origin.copy()

### `RF02` - Inspecionar e Descrever os Dados

Visualizar o  _dataset_ utilizando o método `head()` da biblioteca `pandas`. Assim como tabelar alguns aspectos do _dataset_ para verificar algumas consistências, tipo dos dados entre outro.

In [3]:
df.head(3)

,id_venda,data_venda,id_cliente,nome_cliente,cidade,estado,regiao,id_produto,produto,categoria,quantidade,preco_unitario,desconto,previsao_entrega,data_entrega
0,ORD00001,2025-01-01,Cliente_008,Thales Pereira,Cascavel,PR,Sul,P0005,"Fone de Ouvido Sem Fio TWS, PHILIPS",Áudio,1.0,124.91,0.0327,2025-01-07,2025-01-09
1,ORD00002,2025-01-01,Cliente_011,Julia Ribeiro,Santa Maria,RS,Sul,P0006,Fone de ouvido Sem Fio QCY T27,Áudio,4.0,129.86,0.0477,2025-01-04,2025-01-08
2,ORD00003,2025-01-01,Cliente_019,Gabriela da Paz,Uberlândia,MG,Sudeste,P0002,Samsung Galaxy Tab S6 Lite,Smartphones e Tablets,2.0,1799.01,0.1481,2025-01-05,2025-01-04


**Dicionário dos Dados:**

In [4]:
display(pd.DataFrame(DICIONARIO_DADOS).style.hide(axis="index"))

Coluna,Descrição
id_venda,Identificador único da venda.
data_venda,Data em que a venda foi realizada.
id_cliente,Identificador do cliente associado à venda.
nome_cliente,Nome do cliente.
cidade,Cidade de residência do cliente.
estado,Estado de residência do cliente.
regiao,Região geográfica associada ao cliente.
id_produto,Identificador do produto vendido.
produto,Nome do produto vendido.
categoria,Categoria à qual o produto pertence.


In [5]:
print(f"{"="*6} Dimensões {"="*6}")
linhas, colunas = df.shape
print(f"{"Linhas":<7}: {linhas}\n{"Colunas":<7}: {colunas}")
print("=" * 23)

====== Dimensões ======
Linhas : 5732
Colunas: 15


In [6]:
def resumo_estrutural(df):
    """
    Exibe um resumo estrutural do DataFrame com
    tipos de dados e valores ausentes.
    """
    resumo = pd.DataFrame(
        {
            "Coluna": df.columns,
            "dtype": df.dtypes.astype(str).values,
            "Valores Ausentes": df.isnull().sum().values,
            "% Ausentes": (df.isnull().mean().values * 100),
        }
    )

    display(resumo.style.hide(axis="index").format({"% Ausentes": "{:.2f}%"}))


resumo_estrutural(df)

Coluna,dtype,Valores Ausentes,% Ausentes
id_venda,str,0,0.00%
data_venda,str,63,1.10%
id_cliente,str,0,0.00%
nome_cliente,str,0,0.00%
cidade,str,0,0.00%
estado,str,0,0.00%
regiao,str,0,0.00%
id_produto,str,0,0.00%
produto,str,0,0.00%
categoria,str,0,0.00%


Os dados estão divididos em dois tipos `str` (strings/textos) e `float64` (números de pontos flutuantes, decimais). O que chama a atenção são colunas de datas (`data_venda`, `previsao_entrega` e `data_entrega`) com tipo `str`.

<fieldset style="
  border: 1px solid #90CDF4;
  border-radius: 10px;
  padding: 18px 22px;
  margin: 16px 0;
  background-color: #EBF8FF;
  box-shadow: 0 4px 10px rgba(49, 130, 206, 0.12);
">
  <legend style="
    font-weight: 600;
    font-size: 1.05em;
    color: #2B6CB0;
    background-color: #EBF8FF;
    padding: 4px 12px;
    border-radius: 12px;
    border: 1px solid #90CDF4;
  ">Observação</legend>
  <div style="
    text-align: justify;
    line-height: 1.6;
  ">
É necessário fazer um parse das colunas de data para o tipo `datetime`. Passos posteriores esse parse será executado.
  </div>
</fieldset>

As colunas `data_venda`, `quantidade` e `preco_unitário` apresentam dados ausentes. Na etapa de tratamento dos dados poderá sem imputado dados respeitando uma lógica ou até mesmo a exclusão de registros que apresentem dados ausentes (exclusão é um passo "mais" seguro pois a soma de registros com dados ausentes é de ~6.5% do total).

## 2. Limpeza e Tratamento dos Dados

### RF03 — Limpar e Tratar os Dados

#### **1. Colunas textuais, identificadores e datas**

Nesta etapa serão avaliadas as colunas armazenadas inicialmente como texto, considerando a natureza de cada variável e os possíveis problemas de qualidade presentes nos dados.

Serão verificados, entre outros aspectos:
- espaços excedentes;
- inconsistências de escrita;
- identificadores fora do padrão esperado;
- caracteres ou separadores inadequados;
- datas inválidas ou que necessitem conversão para um tipo temporal apropriado.

<div style="
  margin: 12px 0;
  padding: 14px 18px;
  border: 1px solid #B794F4;
  border-left: 4px solid #805AD5;
  border-radius: 8px;
  background-color: #FAF5FF;
  color: #44337A;
  box-shadow: 0 3px 8px rgba(107, 70, 193, 0.12);
">
  <div style="margin: 0;">

As colunas `nome_cliente`, `cidade`, `estado` e `regiao` não apresentam valores ausentes, espaços excedentes ou inconsistências textuais identificáveis nesta etapa. Esse comportamento é coerente com a forma como esses atributos foram controlados durante a preparação do dataset.

As colunas textuais `id_venda`, `id_cliente` e `id_produto` passarão por verificações específicas de integridade, incluindo valores ausentes, espaços excedentes e aderência aos respectivos padrões de identificação. As colunas `produto` e `categoria` também serão avaliadas quanto à consistência e padronização textual.

As colunas relacionadas a datas serão convertidas para o tipo `datetime`, utilizando a própria conversão como mecanismo de validação para identificar valores ausentes ou incompatíveis com uma representação temporal válida.
  </div>
</div>


In [7]:
registros_iniciais = len(df)

**`id_venda`, `id_cliente` e `id_produto`**

Os identificadores possuem padrões esperados específicos:

- `id_venda`: `ORD` seguido de cinco dígitos;
- `id_cliente`: `Cliente_` seguido de três dígitos;
- `id_produto`: `P` seguido de quatro dígitos.

Serão verificados valores ausentes, aderência ao padrão esperado e, no caso de `id_venda`, possíveis duplicidades.

In [8]:
colunas = "id_venda id_cliente id_produto".split()

padroes = [
    r"^ORD\d{5}$",
    r"^Cliente_\d{3}$",
    r"^P\d{4}$",
]


def verificar_padrao(df, col, pad):
    print(f"{'=' * 8} {col:^12} {'=' * 8}")

    validos = df[col].str.fullmatch(pad, na=False)

    print("Padrão válido:", validos.all())
    print("Valores ausentes:", df[col].isna().sum())
    print("Fora do padrão:", (~validos).sum())

    if col == "id_venda":
        print("Valores duplicados:", df[col].duplicated().sum())

    print("=" * 30 + "\n")


for col, pad in zip(colunas, padroes):
    verificar_padrao(df, col, pad)

========   id_venda   ========
Padrão válido: True
Valores ausentes: 0
Fora do padrão: 0
Valores duplicados: 0

========  id_cliente  ========
Padrão válido: False
Valores ausentes: 0
Fora do padrão: 226

========  id_produto  ========
Padrão válido: True
Valores ausentes: 0
Fora do padrão: 0



> `id_cliente` apresenta 226 registros fora do padrão esperado. Esses registros serão normalizados com expressão regular e submetidos novamente à validação.

In [9]:
# Limpeza com re.sub() e .strip()
padrao_cliente = padroes[1]


def normalizar_id_clientes(valor):
    match = re.search(r"cliente\D*(\d{3})", str(valor).strip(), flags=re.IGNORECASE)

    if match:
        return f"Cliente_{match.group(1)}"

    return valor


df["id_cliente"] = df.id_cliente.apply(normalizar_id_clientes)

verificar_padrao(df, "id_cliente", padrao_cliente)

========  id_cliente  ========
Padrão válido: True
Valores ausentes: 0
Fora do padrão: 0



In [10]:
for col in ["produto", "categoria"]:
    original = df[col].copy()
    df[col] = original.str.strip()

    print(f"{'=' * 8} {col:^12} {'=' * 8}")
    print(f"Únicos originais:           {original.nunique()}")
    print(f"Únicos após strip: {df[col].nunique()}")
    print(f"Com espaços excedentes:     {original.ne(original.str.strip()).sum()}")
    print()

========   produto    ========
Únicos originais:           72
Únicos após strip: 27
Com espaços excedentes:     103

========  categoria   ========
Únicos originais:           15
Únicos após strip: 5
Com espaços excedentes:     69



In [11]:
print("=" * 8 + " Lista de Produtos " + "=" * 8 + "\n")
for i, valor in enumerate(sorted(df["produto"].dropna().unique()), start=1):
    print(f"{i:02d}. {valor}")

print("\n" + "=" * 8 + " Lista de Categorias " + "=" * 8 + "\n")
for i, valor in enumerate(sorted(df["categoria"].dropna().unique()), start=1):
    print(f"{i:02d}. {valor}")

======== Lista de Produtos ========

01. ACER Notebook Gamer Nitro
02. Cabo USB-C Trançado 2m
03. Caixa de Som Bluetooth JBL
04. Carregador Turbo Tipo-C 50w
05. Carregador Turbo USB-C 30W
06. Fone de Ouvido Profissional AKG K92
07. Fone de Ouvido Profissional EDIFIER WH950NB
08. Fone de Ouvido Sem Fio TWS, PHILIPS
09. Fone de ouvido Sem Fio QCY T27
10. Memória Kingston Fury 16GB DDR5
11. Monitor UHD Samsung 32"
12. Mouse Redragon Wireless
13. Película de Vidro 3D Galaxy A30s
14. Película para iPhone 14Pro Anti impacto
15. Projetor EPSON Powerlite Wide Screen
16. Projetor Smart Epson EpiqVision, FULL HD
17. SSD Crucial 1TB NVMe M.2
18. Samsung Galaxy A36
19. Samsung Galaxy Tab S6 Lite
20. Smart TV 32” Philco LED Roku TV
21. Smart TV LED 32"Samsung LED Tizen TV
22. Smartphone Motorola Edge 50 Neo
23. Soundbar Samsung com 2.1 canais, Dolby Digital
24. Soundbar Tomate MTS-2017, Bivolt
25. Tablet Lenovo Tab M11
26. Teclado Mecanico Fizz
27. Webcam Logitech Full HD

======== Lista de Categor

**`data_venda`, `previsao_entrega` e `data_entrega`**

In [12]:
colunas_datas = [
    "data_venda",
    "previsao_entrega",
    "data_entrega",
]

datas_convertidas = df[colunas_datas].apply(pd.to_datetime, errors="coerce")

In [13]:
for col in colunas_datas:
    falhas = datas_convertidas[col].isna() & df[col].notna()

    print(f"{'=' * 8} {col:^18} {'=' * 8}")
    print("Falhas na conversão:", falhas.sum())
    print("Todos convertidos:  ", not falhas.any())
    print()

========     data_venda     ========
Falhas na conversão: 0
Todos convertidos:   True

========  previsao_entrega  ========
Falhas na conversão: 0
Todos convertidos:   True

========    data_entrega    ========
Falhas na conversão: 0
Todos convertidos:   True



As três colunas temporais foram submetidas à conversão com
`pd.to_datetime(..., errors="coerce")`.

Não foram identificadas falhas de conversão entre os valores preenchidos,
indicando que os formatos existentes são reconhecidos pelo Pandas.
Entretanto, `data_venda` possui registros originalmente ausentes. Como essa
variável é necessária para as análises temporais posteriores, esses registros
serão removidos.

In [14]:
df[colunas_datas] = datas_convertidas

In [15]:
print(df[colunas_datas].dtypes)

data_venda          datetime64[us]
previsao_entrega    datetime64[us]
data_entrega        datetime64[us]
dtype: object


Para os valores de `data_venda` ausentes optei por excluir os registros utilizando o método `.dropna()`

In [16]:
removidos_data = df["data_venda"].isna().sum()

df = df.dropna(subset=["data_venda"]).copy()

In [17]:
resumo_estrutural(df)

Coluna,dtype,Valores Ausentes,% Ausentes
id_venda,str,0,0.00%
data_venda,datetime64[us],0,0.00%
id_cliente,str,0,0.00%
nome_cliente,str,0,0.00%
cidade,str,0,0.00%
estado,str,0,0.00%
regiao,str,0,0.00%
id_produto,str,0,0.00%
produto,str,0,0.00%
categoria,str,0,0.00%


#### **2. Colunas numéricas**

**Quantidade**

- verificar valores ausentes;
- verificar valores negativos ou iguais a zero;
- confirmar se os valores existentes são inteiros;
- remover registros com quantidade ausente;
- converter o tipo para `int64`;
- validar novamente.

In [18]:
print(f"{'=' * 8} {'quantidade':^16} {'=' * 8}")

print("dtype inicial:", df["quantidade"].dtype)

removidos_quantidade = df["quantidade"].isna().sum()

print("Valores ausentes:", removidos_quantidade)

if removidos_quantidade > 0:
    df = df.dropna(subset=["quantidade"]).copy()

todos_inteiros = (df["quantidade"] % 1 == 0).all()
print("Todos os valores são inteiros:", todos_inteiros)

if todos_inteiros:
    df["quantidade"] = df["quantidade"].astype("int64")

print("dtype final:", df["quantidade"].dtype)
print("Valor mínimo:", df["quantidade"].min())
print("Valor máximo:", df["quantidade"].max())

========    quantidade    ========
dtype inicial: float64
Valores ausentes: 228
Todos os valores são inteiros: True
dtype final: int64
Valor mínimo: 1
Valor máximo: 4


**Preço unitário**

- verificar valores ausentes;
- verificar valores negativos ou iguais a zero;
- observar coerência dos preços existentes;
- remover registros com preço ausente;
- garantir o tipo `float64`;
- validar novamente.


In [19]:
print(f"{'=' * 8} {'preco_unitario':^16} {'=' * 8}")

print("dtype inicial:", df["preco_unitario"].dtype)

removidos_preco = df["preco_unitario"].isna().sum()

print("Valores ausentes:", removidos_preco)

if removidos_preco > 0:
    df = df.dropna(subset=["preco_unitario"]).copy()

precos_positivos = (df["preco_unitario"] > 0).all()

print("Todos os preços são positivos:", precos_positivos)
print("Valor mínimo:", df["preco_unitario"].min())
print("Valor máximo:", df["preco_unitario"].max())
print("dtype final:", df["preco_unitario"].dtype)

========  preco_unitario  ========
dtype inicial: float64
Valores ausentes: 80
Todos os preços são positivos: True
Valor mínimo: 17.9
Valor máximo: 4604.0
dtype final: float64


**Desconto**

- verificar valores ausentes;
- verificar valores negativos;
- verificar se os valores permanecem dentro da faixa definida para desconto;
- garantir tipo numérico adequado;
- validar novamente.

In [20]:
print(f"{'=' * 8} {'desconto':^16} {'=' * 8}")

print("dtype inicial:", df["desconto"].dtype)

valores_nulos = df["desconto"].isna().sum()
print("Valores ausentes:", valores_nulos)

descontos_validos = df["desconto"].dropna().between(0, 1).all()

print("Todos os descontos estão entre 0 e 1:", descontos_validos)
print("Valor mínimo:", df["desconto"].min())
print("Valor máximo:", df["desconto"].max())
print("dtype final:", df["desconto"].dtype)

========     desconto     ========
dtype inicial: float64
Valores ausentes: 0
Todos os descontos estão entre 0 e 1: True
Valor mínimo: 0.0001
Valor máximo: 0.35
dtype final: float64


**Resultado da limpeza**

Após a padronização das colunas textuais, conversão das datas, normalização
dos identificadores e remoção dos registros com ausência nas variáveis
críticas, o dataset está estruturalmente preparado para a etapa de
transformação e criação das variáveis derivadas do RF04.

In [21]:
registros_finais = len(df)
total_removidos = registros_iniciais - registros_finais

print("=" * 8 + " Relatório de Limpeza " + "=" * 8)

print(f"Registros iniciais:             {registros_iniciais}")
print(f"Removidos por data:             {removidos_data}")
print(f"Removidos por quantidade:       {removidos_quantidade}")
print(f"Removidos por preço unitário:   {removidos_preco}")
print(f"Total de registros removidos:   {total_removidos}")
print(f"Registros finais:               {registros_finais}")

======== Relatório de Limpeza ========
Registros iniciais:             5732
Removidos por data:             63
Removidos por quantidade:       228
Removidos por preço unitário:   80
Total de registros removidos:   371
Registros finais:               5361


In [22]:
resumo_estrutural(df)

Coluna,dtype,Valores Ausentes,% Ausentes
id_venda,str,0,0.00%
data_venda,datetime64[us],0,0.00%
id_cliente,str,0,0.00%
nome_cliente,str,0,0.00%
cidade,str,0,0.00%
estado,str,0,0.00%
regiao,str,0,0.00%
id_produto,str,0,0.00%
produto,str,0,0.00%
categoria,str,0,0.00%


## 3. Transformação e Criação de Colunas Derivadas

### `RF04` - Criar Colunas Derivadas com Transformações Condicionais

#### 1. **`receita_total` & `valor_desconto` & `faixa_receita_item`**

Como cada registro é considerado como uma venda efetuada e entregue, vou adicionar uma nova coluna `receita_total` = `quantidade` * (`preco_unitario` - (`preco_unitario` * `desconto`))

In [23]:
receita_total = df.quantidade * (
    df.preco_unitario - (df.preco_unitario * df.desconto)
).round(2)

df["receita_total"] = receita_total

valor_desconto = (df.preco_unitario * df.quantidade) - df.receita_total

df["valor_desconto"] = valor_desconto

Nesta etapa, a coluna `receita_total` será utilizada para criar uma classificação categórica do valor de cada venda.

A nova coluna `faixa_receita_item` dividirá as transações em três grupos:

- **Baixo Valor:** receita inferior a R\$ 500,00;
- **Médio Valor:** receita entre R\$ 500,00 e R\$ 4.999,99;
- **Alto Valor:** receita igual ou superior a R\$ 5.000,00.

A classificação será feita com `np.select`, permitindo aplicar múltiplas condições de forma vetorizada sobre o DataFrame, sem a necessidade de percorrer os registros com laços.


In [24]:
condicoes = [
    df.receita_total < 500,
    (df.receita_total >= 500) & (df.receita_total < 5000),
    df.receita_total >= 5000,
]

faixas = ["Baixo Valor", "Médio Valor", "Alto Valor"]

df["faixa_receita_item"] = np.select(condicoes, faixas, default="Não Classificado")

#### 2. **`mes_venda` & `bimestre_venda`**

- determinar intervalo temporal
- gerar coluna com nome do mês `mes_venda`

In [25]:
# Detreminar intervalo temporal
primeiro_mes = df.data_venda.min().month
ultimo_mes = df.data_venda.max().month

print(
    f"""
O intervalo de vendas é compreendido entre os mês {primeiro_mes} até mês {ultimo_mes}
"""
)


O intervalo de vendas é compreendido entre os mês 1 até mês 6



Como o dataset compreende um intervalo temporal que vai do mês 1 (Janeiro) até o mês 6 (Junho). Gerar coluna para `bimestre_venda`:

- `B1` Janeiro e Fevereiro
- `B2` Março e Abril
- `B3` Maio e Junho

In [26]:
meses = {1: "Janeiro", 2: "Fevereiro", 3: "Março", 4: "Abril", 5: "Maio", 6: "Junho"}
bimestres = {1: "B1", 2: "B1", 3: "B2", 4: "B2", 5: "B3", 6: "B3"}

df["mes_venda"] = df["data_venda"].dt.month.map(meses)

df["bimestre"] = df["data_venda"].dt.month.map(bimestres)

#### 3. **`desvio_entrega_dias` & `atrasado`**

Nesta etapa, serão criadas duas colunas derivadas a partir de `previsao_entrega` e `data_entrega`.

A coluna `desvio_entrega_dias` representará a diferença, em dias, entre a entrega realizada e a previsão:

* valor negativo: entrega antecipada;
* valor zero: entrega no prazo;
* valor positivo: entrega atrasada.

Também será criada a coluna booleana `atrasado`, indicando diretamente se a entrega ocorreu após a data prevista.

Essas duas colunas permitem analisar tanto a intensidade do desvio quanto a ocorrência de atraso.


In [27]:
df["desvio_entrega_dias"] = (df.data_entrega - df.previsao_entrega).dt.days

df["atrasado"] = df.desvio_entrega_dias > 0

#### Organizar colunas e salvar `vendas_processado.csv`

**Organizar Colunas**

In [28]:
ordem_colunas = [
    "id_venda",
    "data_venda",
    "mes_venda",
    "bimestre",
    "id_cliente",
    "nome_cliente",
    "cidade",
    "estado",
    "regiao",
    "id_produto",
    "produto",
    "categoria",
    "quantidade",
    "preco_unitario",
    "receita_total",
    "faixa_receita_item",
    "desconto",
    "valor_desconto",
    "previsao_entrega",
    "data_entrega",
    "desvio_entrega_dias",
    "atrasado",
]

df = df[ordem_colunas]

In [29]:
print(f"{"="*6} Novas Dimensões {"="*6}")
linhas, colunas = df.shape
print(f"{"Linhas":<7}: {linhas}\n{"Colunas":<7}: {colunas}")
print("=" * 23)

====== Novas Dimensões ======
Linhas : 5361
Colunas: 22


Desde o dataset bruto até esta etapa de processamento, foram adicionadas **7 novas colunas derivadas**, construídas a partir das informações já existentes. Essas variáveis ampliam o conjunto original com indicadores temporais, financeiros, logísticos e categóricos, preparando o DataFrame para as próximas etapas de análise e agregação.


In [30]:
pasta_processed = Path("../data/processed")
pasta_processed.mkdir(parents=True, exist_ok=True)

caminho_saida = pasta_processed / "vendas_processado.csv"

df.to_csv(caminho_saida, index=False, encoding="utf-8-sig")

print(f"CSV salvo - Pasta `data/processed`")

CSV salvo - Pasta `data/processed`


## 4. Análise e Agregação das Métricas de Vendas

Nas etapas anteriores, os dados de vendas foram carregados, inspecionados, limpos e transformados. As inconsistências relevantes foram tratadas e novas variáveis foram criadas para representar informações financeiras, temporais, categóricas e logísticas.

Com isso, o DataFrame processado deixa de ser apenas um conjunto de registros individuais e passa a constituir uma base preparada para análise. A partir desta etapa, o objetivo é reunir as vendas em grupos capazes de responder a questões relacionadas ao desempenho comercial do período.

### `RF05` — Calcular Métricas Agregadas com `groupby`

Cada linha do DataFrame representa uma venda individual. Entretanto, a observação isolada dessas transações não permite identificar diretamente tendências ou comparar o desempenho entre diferentes dimensões do negócio.

Para transformar os registros em informações analíticas, serão utilizadas operações de agrupamento com o método `groupby()` do Pandas. As vendas serão consolidadas para responder às seguintes questões:

* Como a receita, a quantidade vendida e o número de vendas se distribuem entre os meses?
* Quais são os cinco produtos que mais geraram receita?
* Qual foi a receita total obtida por cada categoria?
* Como as regiões se comparam em receita total e ticket médio?

Cada agrupamento resultará em um novo DataFrame, organizado em formato tabular. Ao final, essas tabelas serão reunidas em um dicionário, formando uma estrutura reutilizável para apresentação dos resultados e para as próximas etapas de visualização e exportação.

In [31]:
# atribuir o novo DataFrame
df = pd.read_csv("../data/processed/vendas_processado.csv")
df.head(2)

,id_venda,data_venda,mes_venda,bimestre,id_cliente,nome_cliente,cidade,estado,regiao,id_produto,...,quantidade,preco_unitario,receita_total,faixa_receita_item,desconto,valor_desconto,previsao_entrega,data_entrega,desvio_entrega_dias,atrasado
0,ORD00001,2025-01-01,Janeiro,B1,Cliente_008,Thales Pereira,Cascavel,PR,Sul,P0005,...,1,124.91,120.83,Baixo Valor,0.0327,4.08,2025-01-07,2025-01-09,2,True
1,ORD00002,2025-01-01,Janeiro,B1,Cliente_011,Julia Ribeiro,Santa Maria,RS,Sul,P0006,...,4,129.86,494.68,Baixo Valor,0.0477,24.76,2025-01-04,2025-01-08,4,True


**Como a receita, a quantidade vendida e o número de vendas se distribuem entre os meses?**

In [32]:
ordem_meses = [
    "Janeiro",
    "Fevereiro",
    "Março",
    "Abril",
    "Maio",
    "Junho",
]

In [33]:
relatorio_mensal = (
    df.groupby("mes_venda")
    .agg(
        receita=("receita_total", "sum"),
        unidades_vendidas=("quantidade", "sum"),
        numero_vendas=("id_venda", "nunique"),
    )
    .reset_index()
)

relatorio_mensal["mes_venda"] = pd.Categorical(
    relatorio_mensal["mes_venda"], categories=ordem_meses, ordered=True
)

relatorio_mensal = relatorio_mensal.sort_values("mes_venda").reset_index(drop=True)

display(relatorio_mensal.style.hide(axis="index").format({"receita": "{:.2f}"}))

mes_venda,receita,unidades_vendidas,numero_vendas
Janeiro,1299266.31,1968,824
Fevereiro,1591691.83,2063,855
Março,2291715.46,2443,991
Abril,2004674.61,2241,900
Maio,1613016.68,2081,880
Junho,1537652.30,2131,911


**Quais são os cinco produtos que mais geraram receita?**

In [34]:
top_produto_receita = (
    df.groupby("produto")
    .agg(
        receita_total=("receita_total", "sum"),
    )
    .reset_index()
)

top_produto_receita = top_produto_receita.sort_values(
    "receita_total", ascending=False
).reset_index(drop=True)

top_5 = top_produto_receita.head(5)

for posicao, linha in enumerate(
    top_5.itertuples(index=False),
    start=1,
):
    print(
        f"{posicao:2}º {linha.produto:<40} — " f"Receita: R$ {linha.receita_total:,.2f}"
    )

 1º ACER Notebook Gamer Nitro                — Receita: R$ 1,848,972.94
 2º Projetor Smart Epson EpiqVision, FULL HD — Receita: R$ 1,244,636.36
 3º Samsung Galaxy A36                       — Receita: R$ 850,244.53
 4º Samsung Galaxy Tab S6 Lite               — Receita: R$ 839,934.77
 5º Projetor EPSON Powerlite Wide Screen     — Receita: R$ 801,184.04


**Qual foi a receita total obtida por cada categoria?**

In [35]:
receita_por_categoria = df.groupby("categoria").agg(
    receita_total=("receita_total", "sum")
)

display(receita_por_categoria.sort_values("receita_total", ascending=False))

,receita_total
categoria,
Informática,3049354.76
TV e Projeção,3022977.33
Áudio,2140350.33
Smartphones e Tablets,2017909.85
Acessórios Mobile,107424.92


**Como as regiões se comparam em receita total e ticket médio?**

In [36]:
relatorio_regional = df.groupby("regiao", as_index=False).agg(
    receita_total=("receita_total", "sum"),
    numero_vendas=("id_venda", "nunique"),
)

relatorio_regional["ticket_medio"] = (
    relatorio_regional.receita_total / relatorio_regional.numero_vendas
).round(2)

relatorio_regional.sort_values("ticket_medio", ascending=False)

,regiao,receita_total,numero_vendas,ticket_medio
2,Sudeste,2829006.77,1435,1971.43
1,Norte,2393107.36,1232,1942.46
0,Nordeste,2339772.12,1217,1922.57
3,Sul,2776130.94,1477,1879.57


#### Métricas Gerais

In [37]:
metricas_gerais = {
    "relatorio_mensal": relatorio_mensal,
    "top_produtos": top_produto_receita,
    "receita_categoria": receita_por_categoria,
    "relatorio_regional": relatorio_regional,
}

### `RF06` — Segmentar Clientes por Nível de Gasto

Para identificar os clientes de maior valor, as vendas serão agrupadas por cliente e a receita de suas transações será acumulada. Em seguida, cada cliente será classificado de acordo com seu gasto total.

| Gasto total acumulado         | Segmento |
| ----------------------------- | -------- |
| Abaixo de R\$ 5.000,00         | Bronze   |
| De R\$ 5.000,00 a R\$ 15.000,00 | Prata    |
| Acima de R\$ 15.000,00         | Ouro     |

A classificação será aplicada com uma função `lambda`, resultando em uma estrutura com o identificador, o nome, o gasto total e o segmento de cada cliente.


In [38]:
segmentacao_cliente = df.groupby(["id_cliente", "nome_cliente"], as_index=False).agg(
    gasto_total=("receita_total", "sum")
)

segmentacao_cliente["segmento"] = segmentacao_cliente.gasto_total.apply(
    lambda gasto: (
        "Bronze" if gasto < 5_000 else "Prata" if gasto <= 15_000 else "Ouro"
    )
)

segmentacao_cliente = segmentacao_cliente.sort_values(
    "gasto_total", ascending=False
).reset_index(drop=True)

#### Clientes com maior gasto acumulado

A segmentação completa será ordenada pelo gasto total em ordem decrescente. A tabela seguinte apresenta os dez clientes que mais geraram receita no período analisado.


In [39]:
top_10_clientes = segmentacao_cliente.head(10)

display(top_10_clientes.style.hide(axis="index").format({"gasto_total": "R$ {:,.2f}"}))

id_cliente,nome_cliente,gasto_total,segmento
Cliente_371,Ana Luiza Nunes,"R$ 116,525.85",Ouro
Cliente_165,Maria Fernanda Teixeira,"R$ 111,976.65",Ouro
Cliente_394,Eduardo Oliveira,"R$ 108,436.37",Ouro
Cliente_008,Thales Pereira,"R$ 96,223.86",Ouro
Cliente_360,Mariane Campos,"R$ 92,852.62",Ouro
Cliente_300,Dra. Maria Sophia Nogueira,"R$ 89,455.87",Ouro
Cliente_218,Isabel Pereira,"R$ 89,303.07",Ouro
Cliente_337,Benício Fernandes,"R$ 87,821.42",Ouro
Cliente_266,Caroline Carvalho,"R$ 87,502.24",Ouro
Cliente_357,Amanda Araújo,"R$ 81,303.37",Ouro


#### Distribuição dos clientes por segmento

Além do ranking individual, a contagem de clientes em cada segmento permite observar como a base está distribuída entre as três faixas de gasto.


In [40]:
distribuicao_segmentos = (
    segmentacao_cliente["segmento"]
    .value_counts()
    .rename_axis("segmento")
    .reset_index(name="numero_clientes")
)

display(distribuicao_segmentos.style.hide(axis="index"))

segmento,numero_clientes
Ouro,245
Prata,182
Bronze,107


#### Síntese da segmentação

Foram classificados 534 clientes: 245 no segmento Ouro, 182 no segmento Prata e 107 no segmento Bronze. O segmento Ouro representa a maior parcela da base, com aproximadamente 45,9% dos clientes.

Ana Luiza Nunes apresentou o maior gasto acumulado, com R$ 116.525,85. Todos os clientes do Top 10 pertencem ao segmento Ouro, evidenciando a concentração dos maiores valores na faixa superior da classificação.

## 5. Operações Numéricas com NumPy

### `RF07` — Aplicar Operações Vetorizadas e Broadcasting

Nesta etapa, a coluna `receita_total` será convertida de uma `Series` do Pandas para um array NumPy. Essa estrutura permitirá aplicar operações numéricas diretamente sobre todos os valores, sem percorrer individualmente cada elemento.

O array será utilizado para calcular estatísticas gerais, realizar uma transformação vetorizada com broadcasting e filtrar as vendas que apresentam receita acima da média.


In [41]:
receitas_array = df.receita_total.to_numpy()

In [42]:
# Tipo dos dados
receitas_array.dtype

dtype('float64')

In [43]:
receitas_array.ndim

1

In [44]:
receitas_array.size

5361

#### Estatísticas gerais da receita

A média, a mediana e o desvio-padrão serão calculados diretamente sobre o array de receitas. Os resultados serão armazenados em um dicionário reutilizável e apresentados em formato tabular.


In [45]:
estatisticas_receita = {
    "media": float(np.mean(receitas_array)),
    "mediana": float(np.median(receitas_array)),
    "desvio": float(np.std(receitas_array)),
}

tabela_estatisticas = (
    pd.DataFrame.from_dict(estatisticas_receita, orient="index", columns=["valor"])
    .rename_axis("medida")
    .reset_index()
)

display(tabela_estatisticas.style.hide(axis="index").format({"valor": "R$ {:,.2f}"}))

medida,valor
media,"R$ 1,928.37"
mediana,R$ 979.70
desvio,"R$ 2,530.21"


A receita média por venda foi de R$ 1.928,37, enquanto a mediana foi de R$ 979,70. A diferença entre essas medidas indica que vendas de maior valor elevam a média, enquanto pelo menos metade das transações permanece abaixo de aproximadamente R$ 980,00.

O desvio-padrão de R$ 2.530,21 demonstra uma dispersão elevada entre os valores das vendas.

#### Normalização vetorizada das receitas

As receitas serão normalizadas pelo método Min–Max, que transforma os valores originais para o intervalo entre 0 e 1. A operação será aplicada diretamente sobre o array, sem laços, utilizando o valor mínimo e o valor máximo como escalares.

A aplicação desses escalares a todos os elementos demonstra o broadcasting do NumPy, enquanto o cálculo simultâneo caracteriza a operação vetorizada.


$$x_{\text{normalizado}} = \dfrac{x - x_{\text{min}}}{x_{\text{max}} - x_{\text{min}}}$$

In [46]:
valor_min = receitas_array.min()
valor_max = receitas_array.max()

receitas_normalizads = (receitas_array - valor_min) / (valor_max - valor_min)

print(
    f"Máximo deve ser 1.0: {receitas_normalizads.max()}\nMínimo deve ser 0.0: {receitas_normalizads.min()}"
)

Máximo deve ser 1.0: 1.0
Mínimo deve ser 0.0: 0.0


#### Vendas com receita acima da média

A média calculada anteriormente será utilizada como referência para identificar as vendas de maior valor. A comparação entre o array de receitas e esse valor produzirá uma máscara booleana, utilizada para selecionar e contar somente as transações com receita acima da média.


In [47]:
valor_medio = estatisticas_receita["media"]

mask_booleana = receitas_array > valor_medio

vendas_acima = receitas_array[mask_booleana]

numero_vendas = len(receitas_array)
numero_vendas_acima = len(vendas_acima)
pct_vendas_acima = round((numero_vendas_acima / numero_vendas) * 100, 2)

print(
    f"""
==================   Resultados   ===================
Número de vendas total:                  {numero_vendas}
Número de vendas receita acima da média: {numero_vendas_acima}
Porcentagem de vendas acima da média   : {pct_vendas_acima}%
"""
)


==================   Resultados   ===================
Número de vendas total:                  5361
Número de vendas receita acima da média: 1720
Porcentagem de vendas acima da média   : 32.08%



#### Síntese das operações NumPy

O array analisado contém 5.361 valores do tipo `float64`. A normalização Min–Max transformou as receitas para o intervalo entre 0 e 1, confirmando os limites mínimo e máximo esperados.

A filtragem booleana identificou 1.720 vendas com receita superior à média de R$ 1.928,37, correspondendo a 32,08% das transações. Todo o processamento foi realizado diretamente sobre o array, sem laços, demonstrando agregações, vetorização, broadcasting e seleção por máscara booleana.

## 6. Visualização dos Resultados

### `RF08` — Criar Visualizações com Matplotlib e Seaborn

As métricas calculadas anteriormente são apresentadas graficamente para destacar tendências, diferenças de desempenho e relações entre variáveis.

#### Evolução da Receita Mensal

Como a receita se comportou ao longo do período analisado?

![Receita total por mês](../reports/figures/receita_por_mes.png)

A receita cresceu de janeiro até março, quando atingiu aproximadamente R$ 2,29 milhões. Após o pico, houve redução gradual entre abril e junho.

#### Produtos com Maior Receita

Quais produtos tiveram maior participação na receita total?

![Produtos com maior receita](../reports/figures/top_produtos.png)

O notebook gamer Acer Nitro apresentou a maior receita, com aproximadamente R$ 1,85 milhão, seguido pelo projetor Epson EpiqVision. Os resultados mostram forte concentração da receita em produtos de maior valor unitário.

#### Relação entre Quantidade e Receita

Vendas com maior quantidade geram necessariamente maior receita?

![Relação entre quantidade e receita](../reports/figures/quantidade_vs_receita.png)

A relação positiva entre quantidade e receita é fraca. O valor da transação também depende do preço dos produtos, permitindo receitas elevadas mesmo em vendas com quantidades moderadas.

#### Cidades com Maior Desempenho Comercial

Quais cidades concentram o maior número de vendas e as maiores receitas?

![Cidades com mais vendas e receita](../reports/figures/top_cidades_vendas_receita.png)

Ribeirão Preto liderou os dois rankings, seguida por Cascavel. As diferenças nas demais posições mostram que um maior número de vendas nem sempre resulta em maior receita, pois o valor das transações também influencia o desempenho de cada cidade.
